prow/config/cache.go

```diff
@@ -0,0 +1,388 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/git/v2"
+)
+
+// Caching implementation overview
+//
+// Consider the expensive function prowYAMLGetter(). This function is
+// expensive to compute (it involves invoking a Git client, walking a filesystem
+// path, etc). In our caching implementation, we save results of this function
+// into a cache (named ProwYAMLCache). The point is to avoid doing the expensive
+// GetPresubmits() and GetPostsubmits() calls if possible, which involves
+// walking the repository to collect YAML information, etc.
+//
+// ProwYAMLCache uses an off-the-shelf LRU cache library for the low-level
+// caching implementation, which uses the empty interface for keys and values.
+// The values are what we store in the cache, and to retrieve them, we have to
+// provide a key (which must be a hashable object). We wrap this cache with a
+// single lock, and use an algorithm for a concurrent non-blocking cache to make
+// it both thread-safe and also resistant to so-called cache stampede, where
+// many concurrent threads all attempt to look up the same (missing) key/value
+// pair from the cache (see Alan Donovan and Brian Kernighan, "The Go
+// Programming Language" (Addison-Wesley, 2016), p. 277).
+
+// ProwYAMLCache is the user-facing cache.
+type ProwYAMLCache struct {
+	LRUCache
+}
```

Comment:
nit: Its a bit unclear to me why we have both of these types. `ProwYAMLCache` doesn't seem to add any functionality or type assurances to `LRUCache`. I'd suggest either removing the extra layer or making it valuable by having it handle the type casting needed for the `LRUCache`.

Keeping these separate to keep the types general is appealing to me. This coalescing LRU cache seems generally useful and reusable. We may have additional applications for it in the future as well. :+1:

Reply:
> nit: Its a bit unclear to me why we have both of these types. `ProwYAMLCache` doesn't seem to add any functionality or type assurances to `LRUCache`.

Agreed.

> I'd suggest either removing the extra layer or making it valuable by having it handle the type casting needed for the `LRUCache`.

I did the latter.

> Keeping these separate to keep the types general is appealing to me. This coalescing LRU cache seems generally useful and reusable. We may have additional applications for it in the future as well. 👍

I agree, and to keep things cleaner, moved out the lower-level `LRUCache` stuff to its own separate package.